File: CSV2HTML_by_Python_Skeleton/csv2html/translator.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""CSV2HTML：総理大臣と徳川幕府の情報「CSV」からWebページ「HTML」を生成。"""
__author__ = 'AOKI Atsushi'
__version__ = '1.0.7'
__date__ = '2021/01/10 (Created: 2016/01/01)'

import datetime
# import locale
import os
import os.path
# import re
import subprocess

# from PIL import Image

from csv2html.downloader import Downloader
#from csv2html.io import IO
from csv2html.table import Table
from csv2html.tuple import Tuple
from csv2html.writer import Writer

# pylint: disable=R0201
# R0201: Method could be a function (no-self-use)

class Translator:
	"""トランスレータ：CSVファイルをHTMLページへと変換するプログラム。"""

	def __init__(self, classOfAttributes):
		"""トランスレータのコンストラクタ。"""

		super().__init__()
		classOfAttributes.flush_base_directory()
		self._input_table = Table('input', classOfAttributes)
		self._output_table = Table('output', classOfAttributes)

	def compute_string_of_days(self, period):
		"""在位日数を計算して、それを文字列にして応答する。"""

		start, end = period.split('〜')
		start = datetime.datetime.strptime(start, '%Y年%m月%d日')
		if end == '':
			end = datetime.datetime.now()
		else:
			end = datetime.datetime.strptime(end, '%Y年%m月%d日')

		return f'{(end - start + datetime.timedelta(days=1)).days:,}'

	def compute_string_of_image(self, a_tuple):
		"""サムネイル画像から画像へ飛ぶためのHTML文字列を作成して、それを応答する。"""

		image_name = a_tuple[1][a_tuple[1].find("/")+1:]
		a_string = ''
		a_string += f'<a name=\"{a_tuple[0]}\" href=\"{a_tuple[1]}\">'
		a_string += f'<img class=\"borderless\" src=\"{a_tuple[2]}\" width=\"25\" height=\"32\" alt=\"{image_name}\">'

		return a_string

	def execute(self):
		"""CSVファイルをHTMLページへと変換する。"""

		# ダウンローダに必要なファイル群をすべてダウンロードしてもらい、
		# 入力となるテーブルを獲得する。
		a_downloader = Downloader(self._input_table)
		a_downloader.perform()

		# トランスレータに入力となるテーブルを渡して変換してもらい、
		# 出力となるテーブルを獲得する。
		print(self._input_table)
		self.translate()
		print(self._output_table)

		# ライタに出力となるテーブルを渡して、
		# Webページを作成してもらう。
		a_writer = Writer(self._output_table)
		a_writer.perform()

		# 作成したページをウェブブラウザで閲覧する。
		class_attributes = self._output_table.attributes().__class__
		base_directory = class_attributes.base_directory()
		index_html = class_attributes.index_html()
		a_command = "open -a 'Safari' " + base_directory + os.sep + index_html
		subprocess.getoutput(a_command)

	@classmethod
	def perform(cls, class_attributes):
		"""属性リストのクラスを受け取り、CSVファイルをHTMLページへと変換する。"""

		# トランスレータのインスタンスを生成する。
		a_translator = cls(class_attributes)
		# トランスレータにCSVファイルをHTMLページへ変換するように依頼する。
		a_translator.execute()



	def translate(self):
		"""CSVファイルを基にしたテーブルから、HTMLページを基にするテーブルに変換する。"""

		if self._input_table.attributes().caption_string() == '総理大臣':
			a_list = []
			a_list.append(self._input_table.attributes().return_names()[self._input_table.attributes().keys().index("no")])
			a_list.append(self._input_table.attributes().return_names()[self._input_table.attributes().keys().index("order")])
			a_list.append(self._input_table.attributes().return_names()[self._input_table.attributes().keys().index("name")])
			a_list.append(self._input_table.attributes().return_names()[self._input_table.attributes().keys().index("kana")])
			a_list.append(self._input_table.attributes().return_names()[self._input_table.attributes().keys().index("period")])
			a_list.append("在位日数")
			a_list.append(self._input_table.attributes().return_names()[self._input_table.attributes().keys().index("school")])
			a_list.append(self._input_table.attributes().return_names()[self._input_table.attributes().keys().index("party")])
			a_list.append(self._input_table.attributes().return_names()[self._input_table.attributes().keys().index("place")])
			a_list.append(self._input_table.attributes().return_names()[self._input_table.attributes().keys().index("image")])
			self._output_table.attributes().names(a_list)

			for aTuple in self._input_table.tuples():
				a_list = []
				a_list.append(aTuple.values()[self._input_table.attributes().keys().index("no")])
				a_list.append(aTuple.values()[self._input_table.attributes().keys().index("order")])
				a_list.append(aTuple.values()[self._input_table.attributes().keys().index("name")])
				a_list.append(aTuple.values()[self._input_table.attributes().keys().index("kana")])
				a_list.append(aTuple.values()[self._input_table.attributes().keys().index("period")])
				a_list.append(self.compute_string_of_days(aTuple.values()[self._output_table.attributes().keys().index("period")]))
				a_list.append(aTuple.values()[self._input_table.attributes().keys().index("school")])
				a_list.append(aTuple.values()[self._input_table.attributes().keys().index("party")])
				a_list.append(aTuple.values()[self._input_table.attributes().keys().index("place")])
				a_tuple = (aTuple.values()[self._input_table.attributes().keys().index("no")], aTuple.values()[self._input_table.attributes().keys().index("image")], aTuple.values()[self._input_table.attributes().keys().index("thumbnail")])
				a_list.append(self.compute_string_of_image(a_tuple))
				self._output_table.add(Tuple(self._output_table.attributes(), a_list))
		else:
			a_list = []
			a_list.append(self._input_table.attributes().return_names()[self._input_table.attributes().keys().index("no")])
			a_list.append(self._input_table.attributes().return_names()[self._input_table.attributes().keys().index("name")])
			a_list.append(self._input_table.attributes().return_names()[self._input_table.attributes().keys().index("kana")])
			a_list.append(self._input_table.attributes().return_names()[self._input_table.attributes().keys().index("period")])
			a_list.append("在位日数")
			a_list.append(self._input_table.attributes().return_names()[self._input_table.attributes().keys().index("family")])
			a_list.append(self._input_table.attributes().return_names()[self._input_table.attributes().keys().index("rank")])
			a_list.append(self._input_table.attributes().return_names()[self._input_table.attributes().keys().index("image")])
			a_list.append(self._input_table.attributes().return_names()[self._input_table.attributes().keys().index("former")])
			a_list.append(self._input_table.attributes().return_names()[self._input_table.attributes().keys().index("cemetery")])
			self._output_table.attributes().names(a_list)

			for aTuple in self._input_table.tuples():
				a_list = []
				a_list.append(aTuple.values()[self._input_table.attributes().keys().index("no")])
				a_list.append(aTuple.values()[self._input_table.attributes().keys().index("name")])
				a_list.append(aTuple.values()[self._input_table.attributes().keys().index("kana")])
				a_list.append(aTuple.values()[self._input_table.attributes().keys().index("period")])
				a_list.append(self.compute_string_of_days(aTuple.values()[self._output_table.attributes().keys().index("period")]))
				a_list.append(aTuple.values()[self._input_table.attributes().keys().index("family")])
				a_list.append(aTuple.values()[self._input_table.attributes().keys().index("rank")])
				a_tuple = (aTuple.values()[self._input_table.attributes().keys().index("no")], aTuple.values()[self._input_table.attributes().keys().index("image")], aTuple.values()[self._input_table.attributes().keys().index("thumbnail")])
				a_list.append(self.compute_string_of_image(a_tuple))
				a_list.append(aTuple.values()[self._input_table.attributes().keys().index("former")])
				a_list.append(aTuple.values()[self._input_table.attributes().keys().index("cemetery")])
				self._output_table.add(Tuple(self._output_table.attributes(), a_list))

File: CSV2HTML_by_Java/csv2html/Translator.java
package csv2html;

import java.util.Arrays;
import java.awt.image.BufferedImage;
import java.io.File;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.HashSet;
import java.util.List;
import java.util.Date;
import java.util.concurrent.TimeUnit;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import javax.imageio.ImageIO;

/**
 * トランスレータ：CSVファイルをHTMLページへと変換するプログラム。
 */
public class Translator extends Object {
	/**
	 * CSVに由来するテーブルを記憶するフィールド。
	 */
	private Table inputTable;

	/**
	 * HTMLに由来するテーブルを記憶するフィールド。
	 */
	private Table outputTable;

	/**
	 * 属性リストのクラスを指定したトランスレータのコンストラクタ。
	 *
	 * @param classOfAttributes 属性リストのクラス
	 */
	public Translator(Class<? extends Attributes> classOfAttributes) {
		super();

		Attributes.flushBaseDirectory();

		try {
			Constructor<? extends Attributes> aConstructor = classOfAttributes.getConstructor(String.class);

			this.inputTable = new Table(aConstructor.newInstance("input"));
			this.outputTable = new Table(aConstructor.newInstance("output"));
		} catch (NoSuchMethodException | InstantiationException | IllegalAccessException
				| InvocationTargetException anException) {
			anException.printStackTrace();
		}

		return;
	}

	/**
	 * 在位日数を計算して、それを文字列にして応答する。
	 *
	 * @param periodString 在位期間の文字列
	 * @return 在位日数の文字列
	 */
	public String computeNumberOfDays(String periodString) {
		Integer datecal;
		List<String> startEnd = new ArrayList<>();
		String[] spritDate = periodString.split("〜");
		for (String temp : spritDate) {
			startEnd.add(temp);
		}
		if (startEnd.size() == 1) {
			Date date = new Date();
			String nowDate = new SimpleDateFormat("yyyy年MM月dd日 hh:mm:ss").format(date);
			startEnd.add(nowDate);
		}
		SimpleDateFormat sdFormat = new SimpleDateFormat("yyyy年MM月dd日");
		try {
			Date date1 = sdFormat.parse(startEnd.get(0));
			Date date2 = sdFormat.parse(startEnd.get(1));
			datecal = Integer.valueOf(
					String.valueOf(TimeUnit.DAYS.convert(date2.getTime() - date1.getTime(), TimeUnit.MILLISECONDS)));
			return String.format("%,d", datecal + 1);
		} catch (ParseException anException) {
			anException.printStackTrace();
		}

		return null;
	}

	/**
	 * サムネイル画像から画像へ飛ぶためのHTML文字列を作成して、それを応答する。
	 *
	 * @param aString 画像の文字列
	 * @param aTuple  タプル
	 * @param no      番号
	 * @return サムネイル画像から画像へ飛ぶためのHTML文字列
	 */
	public String computeStringOfImage(String aString, Tuple aTuple, int no) {
		String anImageString = this.inputTable.attributes().baseDirectory().concat(aString);
		BufferedImage anImage = null;
		try {
			anImage = ImageIO.read(new File(anImageString));
		} catch (Exception anException) {
			anException.printStackTrace();
		}
		String anotherString = "<a name=\"%d\" href=\"%s\" >".formatted(no, aString);
		anotherString = anotherString
				.concat("<img class=\"borderless\" src=\" %s \" width=\"%d\" height=\"%d\" alt=\" %s \">"
						.formatted(aTuple.values().get(this.inputTable.attributes().indexOfThumbnail()),
								anImage.getWidth(), anImage.getHeight(),
								aString.substring(aString.lastIndexOf("/") + 1)));

		return anotherString;
	}

	/**
	 * CSVファイルをHTMLページへ変換する。
	 */
	public void execute() {
		// 必要な情報をダウンロードする。
		Downloader aDownloader = new Downloader(this.inputTable);
		aDownloader.perform();

		// CSVに由来するテーブルをHTMLに由来するテーブルへと変換する。
		System.out.println(this.inputTable);
		this.translate();
		System.out.println(this.outputTable);

		// HTMLに由来するテーブルから書き出す。
		Writer aWriter = new Writer(this.outputTable);
		aWriter.perform();

		// ブラウザを立ち上げて閲覧する。
		try {
			Attributes attributes = this.outputTable.attributes();
			String fileStringOfHTML = attributes.baseDirectory() + attributes.indexHTML();
			ProcessBuilder aProcessBuilder = new ProcessBuilder("open", "-a", "Safari", fileStringOfHTML);
			aProcessBuilder.start();
		} catch (Exception anException) {
			anException.printStackTrace();
		}

		return;
	}

	/**
	 * 属性リストのクラスを受け取って、CSVファイルをHTMLページへと変換するクラスメソッド。
	 *
	 * @param classOfAttributes 属性リストのクラス
	 */
	public static void perform(Class<? extends Attributes> classOfAttributes) {
		// トランスレータのインスタンスを生成する。
		Translator aTranslator = new Translator(classOfAttributes);
		// トランスレータにCSVファイルをHTMLページへ変換するように依頼する。
		aTranslator.execute();

		return;
	}

	/**
	 * CSVファイルを基にしたテーブルから、HTMLページを基にするテーブルに変換する。
	 */
	public void translate() {
		List<String> aList;
		if (this.inputTable.attributes().captionString().equals("総理大臣")) {
			aList = new ArrayList<String>();
			aList.add(this.inputTable.attributes().nameAt(this.inputTable.attributes().indexOfNo()));
			aList.add(this.inputTable.attributes().nameAt(this.inputTable.attributes().indexOf("order")));
			aList.add(this.inputTable.attributes().nameAt(this.inputTable.attributes().indexOfName()));
			aList.add(this.inputTable.attributes().nameAt(this.inputTable.attributes().indexOfKana()));
			aList.add(this.inputTable.attributes().nameAt(this.inputTable.attributes().indexOfPeriod()));
			aList.add("在位日数");
			aList.add(this.inputTable.attributes().nameAt(this.inputTable.attributes().indexOf("school")));
			aList.add(this.inputTable.attributes().nameAt(this.inputTable.attributes().indexOf("party")));
			aList.add(this.inputTable.attributes().nameAt((this.inputTable.attributes().indexOf("place"))));
			aList.add(this.inputTable.attributes().nameAt(this.inputTable.attributes().indexOfImage()));
			this.outputTable.attributes().names(aList);

			for (Tuple aTuple : this.inputTable.tuples()) {
				aList = new ArrayList<String>();
				aList.add(aTuple.values().get(this.inputTable.attributes().indexOfNo()));
				aList.add(aTuple.values().get(this.inputTable.attributes().indexOf("order")));
				aList.add(aTuple.values().get(this.inputTable.attributes().indexOfName()));
				aList.add(aTuple.values().get(this.inputTable.attributes().indexOfKana()));
				aList.add(aTuple.values().get(this.inputTable.attributes().indexOfPeriod()));
				aList.add(this.computeNumberOfDays(aTuple.values().get(this.outputTable.attributes().indexOfPeriod())));
				aList.add(aTuple.values().get(this.inputTable.attributes().indexOf("school")));
				aList.add(aTuple.values().get(this.inputTable.attributes().indexOf("party")));
				aList.add(aTuple.values().get(this.inputTable.attributes().indexOf("place")));
				aList.add(this.computeStringOfImage(
						aTuple.values().get(this.inputTable.attributes().indexOfThumbnail()),
						aTuple, Integer.valueOf(aTuple.values().get(this.inputTable.attributes().indexOfNo()))));
				this.outputTable.add(new Tuple(this.outputTable.attributes(), aList));
			}
		} else {
			aList = new ArrayList<String>();
			aList.add(this.inputTable.attributes().nameAt(this.inputTable.attributes().indexOfNo()));
			aList.add(this.inputTable.attributes().nameAt(this.inputTable.attributes().indexOfName()));
			aList.add(this.inputTable.attributes().nameAt(this.inputTable.attributes().indexOfKana()));
			aList.add(this.inputTable.attributes().nameAt(this.inputTable.attributes().indexOfPeriod()));
			aList.add("在位日数");
			aList.add(this.inputTable.attributes().nameAt(this.inputTable.attributes().indexOf("family")));
			aList.add(this.inputTable.attributes().nameAt(this.inputTable.attributes().indexOf("rank")));
			aList.add(this.inputTable.attributes().nameAt(this.inputTable.attributes().indexOfImage()));
			aList.add(this.inputTable.attributes().nameAt(this.inputTable.attributes().indexOf("former")));
			aList.add(this.inputTable.attributes().nameAt(this.inputTable.attributes().indexOf("cemetery")));
			this.outputTable.attributes().names(aList);

			for (Tuple aTuple : this.inputTable.tuples()) {
				aList = new ArrayList<String>();
				aList.add(aTuple.values().get(this.inputTable.attributes().indexOfNo()));
				aList.add(aTuple.values().get(this.inputTable.attributes().indexOfName()));
				aList.add(aTuple.values().get(this.inputTable.attributes().indexOfKana()));
				aList.add(aTuple.values().get(this.inputTable.attributes().indexOfPeriod()));
				aList.add(this.computeNumberOfDays(aTuple.values().get(this.outputTable.attributes().indexOfPeriod())));
				aList.add(aTuple.values().get(this.inputTable.attributes().indexOf("family")));
				aList.add(aTuple.values().get(this.inputTable.attributes().indexOf("rank")));
				aList.add(this.computeStringOfImage(
						aTuple.values().get(this.inputTable.attributes().indexOfThumbnail()),
						aTuple, Integer.valueOf(aTuple.values().get(this.inputTable.attributes().indexOfNo()))));
				aList.add(aTuple.values().get(this.inputTable.attributes().indexOf("former")));
				aList.add(aTuple.values().get((this.inputTable.attributes().indexOf("cemetery"))));
				this.outputTable.add(new Tuple(this.outputTable.attributes(), aList));
			}
		}

		return;
	}
}
